packages/SwingSet/src/vats/vattp/vat-vattp.js

```diff
@@ -140,51 +184,54 @@ export function buildRootObject(vatPowers) {
       mailbox = mailboxDevnode;
     },
 
-    /*
-     * 'name' is a string, and must match the name you pass to
-     * deliverInboundMessages/deliverInboundAck
+    /**
+     * @param {string} name  Unique name identifying the remote for "deliverInbound" functions
      */
     addRemote(name) {
-      assert(!remotes.has(name), X`already have remote ${name}`);
-      const r = getRemote(name);
+      assert(!mailboxes.has(name), X`already have remote ${name}`);
+      const r = provideMailbox(name);
       const transmitter = Far('transmitter', {
         transmit(msg) {
-          const o = r.outbound;
-          const num = o.highestAdded + 1;
-          // console.debug(`transmit to ${name}[${num}]: ${msg}`);
+          // Durability question: are these references to `r` and `name` safe?
+          // If not, then what is the proper strategy?
+          // Do we need to be durable all the way up, with "vat-tp handler" being
+          // a durable singleton with state referencing `mailboxes` from baggage,
+          // an addRemote function reaching into that state and returning a
+          // durable object of its own, and deliverInboundMessages/deliverInboundAck
+          // functions that also leverage that state?
```

Comment:
yeah, basically that. When we're done, the only `Far` that should appear in this file is the one that builds the root object. Everything else needs to be a durable Kind, or in a durable collection.

`r` can be used safely from durable data, because it's an instance of a durable kind (the `Mailbox` kind). And `name` is durable-safe, since it's just a string.

The approach I'd take is to use `defineKindMulti()` to create a cohort of `{ transmitter, receiverSetter, inbound }` objects for each remote, which would subsume the `Mailbox` kind you defined. The `state` of these objects will include the counters and the `receiver` object reference, and also the `name` of the remote. The  transmitter facet has the `transmit(msg)` method (which does stuff with the counters, then uses the `name` to make the call to `D(mailbox).add`. The receiverSetter facet will update `state.receiver`. The `inbound` object will implement `deliverMessages(newMessages)` and `deliverAck(ack)` for each name, so the shared `deliverInboundAck(name, ack)` will do `inbound.get(name).deliverAck(ack)`, and `deliverAck()` will do a bunch of `state.highestRemoved + 1` -type stuff. `deliverMessages()` will use `E(state.receiver).receive(body)` and `state.highestDelivered = num`.


You'll need a `makeScalarBigMapStore` to map from remote `name` to the `inbound` facet (but feel free to call it a mailbox or a mailbox handler or whatever), since that's the one place where we'll see messages for lots of different remotes all get sent to the same object. (in the outbound direction, each remote uses a distinct `transmitter` object, but since those will share counter state with the inbound side, having `transmitter` be one facet, and the inbound side be a different facet, both of the same cohort and sharing state, seems like a nice design). You won't need a map to remember transmitters: vat-comms will hold onto those object references from afar. And you won't need a map to remember the receiverSetters: we'll return those to vat-bootstrap, which will use it briefly and then drop it. Only the inbound side (which receives messages from the mailbox device) needs to lookup anything using the remote name.